### PR TITLE
add --profile option to run command

### PR DESCRIPTION
**Pull request by agramfort**
cc @PABannier

example of usage using OLS benchmark. The diff:

Screenshot 2021-07-14 at 11 06 10

then the usage:

Screenshot 2021-07-14 at 11 05 59

### CODE REVIEW

**Review by tomMoral — COMMENTED** [review]
Not a big fan of the globals. I propose another way to do it.
Let me know if you feel that I am nitpicking too much, I tend to do that sometime.

It is just that I don't really like the idea of a global dictionary.
Also, I feel this is a bit brittle because if `profile` is imported before the set of the glolbal variable, this can fail. The proposed solution should be robust in any cases as the decision is taken when the decorator is called.

File: benchopt/profiling.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    profile = line_profiler.LineProfiler()
+else:
+    class Profile:
+        def __call__(self, func):
+            return func
+
+        def print_stats(self):
+            return
+
+    profile = Profile()
```
[the remark this answers is not in the document]
Comment (agramfort) [reply benchopt/profiling.py:14]: Works for me. Don’t hesitate to push here.

File: benchopt/runner.py [hunk 2]
```diff
@@ -497,4 +504,8 @@ def run_benchmark(benchmark, solver_names=None, forced_solvers=None,
     if plot_result:
         from benchopt.plotting import plot_benchmark
         plot_benchmark(save_file, benchmark, html=html)
+
+    if do_profile:
+        print_stats()
```
Comment (tomMoral) [inline benchopt/runner.py:509]: You can simply call this as the Profiler object will only be created if `do_profile` is `True` and at least one function as been decorated.
```suggestion
    print_stats()
```

File: benchopt/runner.py [hunk 3]
```diff
@@ -402,6 +406,9 @@ def run_benchmark(benchmark, solver_names=None, forced_solvers=None,
     """
     print("BenchOpt is running")
 
+    if do_profile:
+        use_profile()
+
```
Comment (tomMoral) [inline benchopt/runner.py:411]: Do you need to call this again?

**Review by tomMoral — COMMENTED** [review]
Just nitpicking but I am not sure we need the `do_profile` in the `run_benchmark` function.

**Comment by agramfort** [comment]
Yes I can simplify if we assume benchopt is only used from CLI. I can then remove the do profile parameter. Shall I do this ?

**Comment by tomMoral** [comment]
I would say yes

**Comment by agramfort** [comment]
@tomMoral let me know how you would add a test for this.

File: benchopt/cli/main.py [hunk 4]
```diff
@@ -85,9 +91,14 @@
               "datasets, see the command `benchopt install`.")
 def run(benchmark, solver_names, forced_solvers, dataset_names,
         objective_filters, max_runs, n_repetitions, timeout,
-        plot=True, html=True, pdb=False, env_name='False'):
+        plot=True, html=True, pdb=False, do_profile=False,
+        env_name='False'):
 
     from benchopt.runner import run_benchmark
+    from benchopt.utils.profiling import use_profile
```
[the remark this answers is not in the document]
Comment (tomMoral) [reply benchopt/cli/main.py:99]: WDYT @agramfort ?

File: benchopt/cli/main.py [hunk 5]
```diff
@@ -104,6 +116,10 @@ def run(benchmark, solver_names, forced_solvers, dataset_names,
             max_runs=max_runs, n_repetitions=n_repetitions,
             timeout=timeout, plot_result=plot, html=html, pdb=pdb
         )
+
+        if do_profile:
+            print_stats()
```
Comment (tomMoral) [inline benchopt/cli/main.py:121]: You can also simply do as it will only print stat if they have been recorded.
```suggestion
        print_stats()
```

**Review by tomMoral — COMMENTED** [review]
For the test, I would add the `profile` decorator to one of the  solver in the dummy benchmark and add a test in `test_cli.py` similar to `test_benchopt_run_in_env`, where you check that with `--profile` you have a profiler report and that without it, you don't?

**Comment by agramfort** [comment]
@tomMoral it's a mystery... it seems the USE_PROFILE is reset back to False although it's set to True first....

**Comment by agramfort** [comment]
Merge if happy @tomMoral

**Comment by tomMoral** [comment]
Merged! thx @agramfort !

**Pull request merged**